agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -57,7 +90,7 @@ private Function createFunction() {
     if (functions.isEmpty()) {
       return IDENTITY_FUNCTION;
     }
-    return new ChainFunction(functions.toArray(new Function[functions.size()]));
+    return new CompiledFunction(recycle(this));
```

Comment:
feels a bit weird for recycle() to return anything. I think the array creation is better inlined here, and then recycle() handles the clearing and recycling, and is a void-returning method. Then finally return new CompiledFunction(theArrayConstructedJustNow).